### ISSUE TITLE
Bug: RuleCreator leaks internal meta.docs, meta.type types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#8695]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Playground Link

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUQMoBlgDsYFUbAA2AznAL5wBmUEIcA5AAIwCeYApsQMZTBgwBaToWwwA9AFcCJegG4AUPPYAPSLDhcIWYvCgTC7OAF4UGUfiLEAdACV97AMJR2AQxjQrAd2AwAFhCkAEQguYgAKBHk4DWc3djDNHBUYAEpEKOi4ZxgJKCx0zMyASQATdhxgSmB2KDCsCDK0yMLCxJhkq2c1GAi4EE5iFwBzdlKALjgAIhUXcANJgBo4erLyFIUW8gzosgyyBYyyyhd9GAB5fmAtYgmAbQBdA+j+mBcJ5sySkJuClrLuXiXLQTej0J4tZyaED9LBlEogyG0GFwsHbTIQPw1CYwPTscE7fF9AbDTjvNHRGZzdgTSZnXzLCAAQkmaP2aNYHBBhBcLACMFRhW4vnYIDecAeewOZHWiiAA

### Repro Code

```TypeScript
import { ESLintUtils } from '@typescript-eslint/utils';

export const rule = ESLintUtils.RuleCreator.withoutDocs({
  create(context) {
    return {
      Identifier(node) {
        context.report({ messageId: "example", node });
      }
    }
  },
  defaultOptions: [],
  meta: {
    docs: {
      description: '',
      recommended: 'recommended',
      other: true,
    },
    messages: {
      example: "Oh no!"
    },
    type: 'layout',
    schema: []
  },
});
```

### ESLint Config

```javascript
n/a
```

### tsconfig

```jsonc
n/a
```

### Expected Result

Custom rules in community plugins shouldn't be tied into our types for what goes into a rule's `meta.docs`. Plugins might define their own, potentially different documentation metadata & strategies.

Maybe we should make `meta.docs` type... `Record<string, unknown>` for consumers? Or a type parameter on `RuleCreator` that defaults to `Record<string, unknown>`?

### Actual Result

* Known properties have types set to ours: e.g. `meta.docs.recommended` is `RuleRecommended | undefined` in 7.2.0
    * After #8364 (7.3.0), it'll be `RuleRecommendation | RuleRecommendationAcrossConfigs<Options> | undefined`
* Unknown properties cause a type error: e.g. `meta.docs.other`

### Additional Info

See also:

* https://github.com/eslint/eslint/issues/17842: for a more general approach in ESLint core
* https://github.com/typescript-eslint/typescript-eslint/pull/8364#discussion_r1527017534: noting that #8364's expansion of the `meta.docs.recommended` property will be visible to users

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `7.2.0` |

💖

**Comment by JoshuaKGoldberg** [comment]
#8833 shows that this impacts `meta.type` too.

**Comment by JoshuaKGoldberg** [comment]
#9025 is merged into the `v8` branch. ✅
